File: Assn5.py
'''
Alexander Fielding
Kim-Wilson Ngo
Tajbir Sandhu
CECS 424 Assignment 5

'''
import sys

#Kicks off the operations on the given number
def operatorSum(number, sum):
    result = []
    for i in range(1, len(number) + 1):
        #keeps us from getting double zeroes as a numIntue
        if i == 1 or (number[0] != "0" and i > 1): 
            numInt = int(number[:i])
            sumHelper(number[i:], number[:i], numInt, numInt, result, sum) 
    
    #accounts for negative numbers
    number = int(number) * -1
    number = str(number)       
    for i in range(2, len(number) + 1):
        #keeps us from getting double zeroes as a numIntue
        if numInt != '-' and i == 1 or (number[0] != "0" and i > 1 and numInt != '-'): 
            numInt = int(number[:i])
            sumHelper("-" + number[i:], number[:i], numInt, numInt, result, sum) 
            
    return result

#Helper function that adds to the list of expressions once number is fully processed
def sumHelper(number, subNum, current, last, result, sum):
    if not number:
        if current == sum:
            result.append(subNum)
        return
    #Processes numbers subparts at a time an adds in the operation
    for i in range(1, len(number) + 1):
        numInt = number[:i]
        if numInt != '-' and i == 1 or (i > 1 and number[0] != "0" and numInt != '-'): # prevent "00*" as a numberber
            sumHelper(number[i:], subNum + "+" + numInt, current + int(numInt), int(numInt), result, sum)
            sumHelper(number[i:], subNum + "-" + numInt, current - int(numInt), -int(numInt), result, sum)
            #sumHelper((number[i:]), '-' + subNum + "+" + numInt, current + int(numInt), int(numInt), result, sum)
            #sumHelper((number[i:]), '-' + subNum + "-" + numInt, current - int(numInt), -int(numInt), result, sum)

def main(number, sum):
    l = operatorSum(number, sum)


    print("# of matches: ", len(l))
    for i in range(len(l)):
        print(i+1, ":", l[i], "=", sum)


if __name__ == '__main__':
    main(str(sys.argv[1]), int(sys.argv[2]))
    #main("123456789", 0)
    

    
#main("123456789", 0)

File: Assn5.java
/*
	Alexander Fielding
	Kim-Wilson Ngo
	Tajbir Sandhu
	CECS 424 Assignment 5
*/

import java.util.ArrayList;
import java.util.List;

public class Assn5 {

	public static void main(String[] args)
	{
		System.out.println("test");

		String number = args[0];
		int sum = Integer.parseInt(args[1]);
		//String number = "123456789";
		//int sum = 0;
		List<String> l = operatorSum(number, sum);
		System.out.println("# of matches: " + l.size());
		for(int i = 0; i < l.size(); i++)
		{
			System.out.println((i+1) + ": " + l.get(i) + "= " + sum);
		}
	}

	/**
	 * Kicks off the operations on the given number
	 * @param number
	 * @param sum
	 * @return list of expressions
	 */
	public static List<String> operatorSum(String number, int sum)
	{
		List<String> result = new ArrayList<String>();

		for(int i = 1; i < number.length() + 1; i++)
		{
			if(i == 1 || (Character.toString(number.charAt(0)) != "0" && i > 1))
			{

				int numInt = Integer.parseInt(number.substring(0, i));
				String numString = number.substring(i);
				String numString2 = number.substring(0,i);
				sumHelper(numString, numString2, numInt, numInt, result, sum);
			}
		}
		
		int temp = Integer.parseInt(number) * -1;
		number = temp + "";
		
		for(int i = 2; i < number.length() + 1; i++)
		{
			
			int numInt = Integer.parseInt(number.substring(0,i));
			if(i == 1 || (Character.toString(number.charAt(0)) != "0" && i > 1))
			{
				numInt = Integer.parseInt(number.substring(0,i));
				String numString = number.substring(i);
				String numString2 = number.substring(0,i);
				sumHelper(("-" + numString), numString2, numInt, numInt, result, sum);
			}
		}
		

		return result;
	}

	/**
	 * Helper function that adds to the list of expression once number is fully processed
	 * @param number
	 * @param subNum
	 * @param current
	 * @param last
	 * @param result
	 * @param sum
	 */
	public static void sumHelper(String number, String subNum, int current, int last, List<String> result, int sum)
	{
		if(number.equals(""))
		{
			if(current == sum)
			{
				result.add(subNum);

			}
			return;
		}

		//Processes numbers subparts at a time an adds in the operation
		for(int i = 1; i < number.length() + 1; i++)
		{
			String numInt = number.substring(0, i);
			String numInt2 = number.substring(i);
			if(!numInt.equals("-") && i == 1 || (!numInt.equals("-") && i > 1 && Character.toString(number.charAt(0)) != "0"))
			{
				sumHelper(numInt2, subNum + "+" + numInt, (current + Integer.parseInt(numInt)), Integer.parseInt(numInt), result, sum);
				sumHelper(numInt2, subNum + "-" + numInt, (current - Integer.parseInt(numInt)), (-1*(Integer.parseInt(numInt))), result, sum);
				
			}
		}
	}

}
